validator/client/src/test/java/tech/pegasys/artemis/validator/client/DutySchedulerTest.java

```diff
@@ -0,0 +1,198 @@
+  private final ValidatorDutyFactory dutyFactory = mock(ValidatorDutyFactory.class);
+  private final StubAsyncRunner asyncRunner = new StubAsyncRunner();
+
+  private final DutyScheduler validatorClient =
+      new DutyScheduler(
+          asyncRunner,
+          validatorApiChannel,
+          dutyFactory,
+          Map.of(VALIDATOR1_KEY, validator1, VALIDATOR2_KEY, validator2));
+
+  @BeforeEach
+  public void setUp() {
+    when(validatorApiChannel.getDuties(any(), any()))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(emptyList())));
+    when(dutyFactory.createAttestationProductionDuty(any()))
+        .thenReturn(mock(AttestationProductionDuty.class));
+  }
+
+  @Test
+  public void shouldFetchDutiesForCurrentAndNextEpoch() {
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+
+    verify(validatorApiChannel).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+    verify(validatorApiChannel).getDuties(UnsignedLong.valueOf(2), VALIDATOR_KEYS);
+  }
+
+  @Test
+  public void shouldNotRefetchDutiesWhichHaveAlreadyBeenRetrieved() {
+    when(validatorApiChannel.getDuties(any(), any())).thenReturn(new SafeFuture<>());
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+
+    verify(validatorApiChannel).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+    verify(validatorApiChannel).getDuties(UnsignedLong.valueOf(2), VALIDATOR_KEYS);
+
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.valueOf(2)));
+
+    // Requests the next epoch, but not the current one because we already have that
+    verify(validatorApiChannel).getDuties(UnsignedLong.valueOf(3), VALIDATOR_KEYS);
+    verifyNoMoreInteractions(validatorApiChannel);
+  }
+
+  @Test
+  public void shouldRetryWhenRequestingDutiesFails() {
+    final SafeFuture<Optional<List<ValidatorDuties>>> request1 = new SafeFuture<>();
+    final SafeFuture<Optional<List<ValidatorDuties>>> request2 = new SafeFuture<>();
+    when(validatorApiChannel.getDuties(UnsignedLong.ONE, VALIDATOR_KEYS))
+        .thenReturn(request1)
+        .thenReturn(request2);
+
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+    verify(validatorApiChannel, times(1)).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+
+    request1.completeExceptionally(new RuntimeException("Nope"));
+    assertThat(asyncRunner.hasDelayedActions()).isTrue();
+    asyncRunner.executeQueuedActions();
+
+    // Should retry request
+    verify(validatorApiChannel, times(2)).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+
+    // And not have any more retries scheduled
+    assertThat(asyncRunner.hasDelayedActions()).isFalse();
+  }
+
+  @Test
+  public void shouldScheduleBlockProposalDuty() {
+    final UnsignedLong blockProposerSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        ValidatorDuties.withDuties(
+            VALIDATOR1_KEY, 5, 3, List.of(blockProposerSlot), UnsignedLong.valueOf(7));
```

Comment:
I might have missed this previously, but why does ValidatorDuties keep a list of blockProposerSlots? A validator proposes a block about every 3 weeks.

Reply:
Validators are randomly assigned blocks to produce so there's no restriction on them being asked to produce twice in one epoch.  This matches the Lighthouse API and Prysm are moving to it as well (they've realised this is likely why they miss block sometimes).